files.py

```diff
@@ -97,50 +104,70 @@ def load_layout(name, popups=True, save_converted=True, printing=True):
             if popups:
                 window.app.popup(window.app, "Error loading file!", window.app.info_image, "The layout is not in valid JSON format (the new .lpl extention).\n\nIf this was renamed from a .LPHKlayout file, please change\nthe extention back to .LPHKlayout and try loading again.", "OK")
             raise
-    
+
     return layout
 
 def save_lp_to_layout(name):
     layout = dict()
-    layout["version"] = FILE_VERSION
-    
+
+    has_subs = False
+
     layout["buttons"] = []
     for x in range(9):
         layout["buttons"].append([])
         for y in range(9):
             color = lp_colors.curr_colors[x][y]
             script_text = scripts.buttons[x][y].script_str
-            
+
             layout["buttons"][-1].append({"color": color, "text": script_text})
-    
+
+    for x in scripts.VALID_COMMANDS:               # for all the commands that exist
+        if x.startswith(SUBROUTINE_PREFIX):        # if this command is a subroutine
+            if not has_subs:
+                layout["subroutines"] = []         # only add the key if required
+                has_subs = True
+            cmd = scripts.VALID_COMMANDS[x]        # get the command
+            layout["subroutines"] += [cmd.routine] # add the command to the list (name is embedded in the subroutine)
+
+    if has_subs:                                   # file version depends on the existance of subroutines
+        layout["version"] = FILE_VERSION_SUBS
+    else:
+        layout["version"] = FILE_VERSION
+
     save_layout(layout=layout, name=name)
 
 def load_layout_to_lp(name, popups=True, save_converted=True, preload=None):
     global curr_layout
     global in_error
     global layout_changed_since_load
-    
+
     converted_to_rg = False
-    
+
     scripts.Unbind_all()
-    window.app.draw_canvas()
-    
+    scripts.Unload_all()                         # remove all existing subroutines when you load a new layout
+    window.Redraw(True)
+
     if preload == None:
         layout = load_layout(name, popups=popups, save_converted=save_converted)
     else:
         layout = preload
-        
+
+    # load subroutines before buttons so you don't get errors on buttons using them
+    if "subroutines" in layout:                  # were subroutines saved?

```

Comment:
you should compare versions instead

Reply:
This is just checking if there is a saved subroutine.  Even in the most recent version it is still legal not to have them (the code won't write it that way, but it would be bad practice to assume a newer version always had subroutines).  Writing only layouts with subroutines in the new format is simply to retain as much backward compatibility as possible.  If a newer format is created, and that backward compatibility is lost, then just checking the version will not be sufficient.  I could *also* check the version, or I could handle an exception if there weren't subroutines stored.  I can do it any way you prefer.

Reply:
Change will be checked in shortly...